Graphical Client Issue

```
Hello everyone This is Antonio from the robobulls team I am having a bit of an 
issue. I am transferring our stuff from windows to Linux we managed to get 
everything set up vision system works and so does the graphical client.
I am trying to enhance the graphical client by adding some extra buttons but to 
even consider tackling that I first need to open and compile the graphical 
client in Qt I believe I found the correct source and after a couple of changes 
I got through a few minor details, the thing holding me back now is an error I 
do not know how to fix or get around I tried search engines but there was 
literally no results for my type of error I am hoping someone may have familiar 
knowledge of the situation or Qt the error is CentralWindow.cpp:119: error: 
'UnicodeUTF8' is not a member of 'QApplication' I have looked at the encoder 
and it is set to UTF8 that was one of the suggestions I found online aside from 
that I am stumped hope someone can assist. 

I am using QtCreator 2.62 and Qt 5.01
I have attached the file in question as well it is all files from the GClient 
provided in the SSL download package  
```

Original issue reported on code.google.com by `RoboBu...@gmail.com` on 27 Mar 2013 at 6:01

Attachments:
- [CentralWindow.cpp](https://storage.googleapis.com/google-code-attachments/ssl-vision/issue-24/comment-0/CentralWindow.cpp)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,6 +8,9 @@ if(USE_QT5)
 	if(POLICY CMP0043) # compatibility with CMake 3.0.1
 	    cmake_policy(SET CMP0043 OLD)
 	endif()
+	if(POLICY CMP0071) # compatibility with CMake 3.10.0
+	    cmake_policy(SET CMP0071 OLD)
+	endif()
 else()
 	cmake_minimum_required(VERSION 2.6)
 endif()

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,16 @@
 project(ssl-vision)
-cmake_minimum_required(VERSION 2.6)
+set(USE_QT5 FALSE CACHE BOOL "Use Qt5 instead of Qt4")
+if(USE_QT5)
+	cmake_minimum_required(VERSION 2.8.9)
+	if(POLICY CMP0020)
+	    cmake_policy(SET CMP0020 NEW) # remove if CMake >= 2.8.11 required
+	endif()
+	if(POLICY CMP0043) # compatibility with CMake 3.0.1
+	    cmake_policy(SET CMP0043 OLD)
+	endif()
+else()
+	cmake_minimum_required(VERSION 2.6)
+endif()
 
 #defines
 add_definitions(-D UNIX)
@@ -12,7 +23,6 @@ SET(EXECUTABLE_OUTPUT_PATH ${PROJECT_SOURCE_DIR}/bin)
 
 # Find third-party libraries
 find_package(OpenGL REQUIRED)
-find_package(Qt4 4.8.0 COMPONENTS QtCore QtGui QtOpenGL QtNetwork REQUIRED)
 find_package(Threads REQUIRED)
 
 find_package(Protobuf REQUIRED)
@@ -26,13 +36,35 @@ find_package(PkgConfig REQUIRED)
 pkg_check_modules(Eigen3 REQUIRED eigen3)
 pkg_check_modules(DC1394 REQUIRED libdc1394-2)
 
+if(USE_QT5)
+	set(CMAKE_INCLUDE_CURRENT_DIR ON)
+	set(CMAKE_AUTOMOC ON)
+	find_package(Qt5Core REQUIRED)
+	find_package(Qt5Widgets REQUIRED)
+	#find_package(Qt5Network REQUIRED)
+	find_package(Qt5OpenGL REQUIRED)
+	# replaced by automoc
+	macro(qt4_wrap_cpp VARNAME)
+		set(${VARNAME} "")
+	endmacro()
+	# wrap functions
+	macro(qt4_wrap_ui)
+		qt5_wrap_ui(${ARGN})
+	endmacro()
+	macro(qt4_add_resources)
+		qt5_add_resources(${ARGN})
+	endmacro()
+	set(QT_LIBRARIES "")
+else()
+	find_package(Qt4 4.8.0 COMPONENTS QtCore QtGui QtOpenGL QtNetwork REQUIRED)
+	include ( ${QT_USE_FILE} )
+endif()
+
 include_directories(
 	${PROJECT_BINARY_DIR}
 	${Eigen3_INCLUDE_DIRS}
 	${PROTOBUF_INCLUDE_DIRS}
 )
-
-include ( ${QT_USE_FILE} )
 
 find_package(mvIMPACT QUIET)
 
@@ -216,18 +248,28 @@ set (libs ${QT_LIBRARIES}
 	${mvIMPACT_LIBRARIES}
 	${OpenCV_LIBS}
 	${FLYCAP_LIBS}
-	sslvision
 )
+target_link_libraries(sslvision ${libs})
+if(USE_QT5)
+	qt5_use_modules(sslvision Widgets)
+endif()
+set (libs ${libs} sslvision)
 
 ## build the main app
 set (target vision)
 add_executable(${target} ${UI_SRCS} ${MOC_SRCS} ${RC_SRCS} ${SRCS})
 target_link_libraries(${target} ${libs})
+if(USE_QT5)
+	qt5_use_modules(${target} Widgets OpenGL)
+endif()
 
 ##build non graphical client
 set (client client)
 add_executable(${client} src/client/main.cpp )
 target_link_libraries(${client} ${libs})
+if(USE_QT5)
+	qt5_use_modules(${client} Core)
+endif()
 
 ##build logging client
 set (lclient logClient)
@@ -239,6 +281,9 @@ add_executable(${lclient} ${LCLIENT_MOC_SRCS}
 	src/logClient/LogControl.cpp
 )
 target_link_libraries(${lclient} ${libs})
+if(USE_QT5)
+	qt5_use_modules(${lclient} Widgets OpenGL)
+endif()
 
 ##build graphical client
 set (gclient graphicalClient)
@@ -248,3 +293,6 @@ add_executable(${gclient} ${GCLIENT_MOC_SRCS}
   src/graphicalClient/gltext.cpp
 )
 target_link_libraries(${gclient} ${libs})
+if(USE_QT5)
+	qt5_use_modules(${gclient} Widgets OpenGL)
+endif()

--- a/src/app/gui/icons/icons_gui.qrc
+++ b/src/app/gui/icons/icons_gui.qrc
@@ -75,7 +75,6 @@
         <file>xload.png</file>
         <file>colors.png</file>
         <file>background.png</file>
-        <file>xedit.png</file>
         <file>undo.png</file>
         <file>redo.png</file>
         <file>arrow-left.png</file>

--- a/src/app/gui/mainwindow.ui
+++ b/src/app/gui/mainwindow.ui
@@ -18,9 +18,6 @@
   <property name="windowTitle" >
    <string>SSL-Vision System, released as GPLv3</string>
   </property>
-  <property name="windowIcon" >
-   <iconset/>
-  </property>
   <property name="dockNestingEnabled" >
    <bool>false</bool>
   </property>

--- a/src/app/main.cpp
+++ b/src/app/main.cpp
@@ -20,8 +20,6 @@
 //========================================================================
 
 #include <QApplication>
-#include <QCleanlooksStyle>
-#include <QPlastiqueStyle>
 #include <QString>
 #include "mainwindow.h"
 #include <signal.h>
@@ -112,9 +110,6 @@ int main(int argc, char *argv[])
 
   MainWindow mainWin(start, enforce_affinity);
   mainWinPtr = &mainWin;
-  //if desired, launch a particular style:
-  // app.setStyle(new QPlastiqueStyle());
-  // app.setStyle(new QCleanlooksStyle());
   mainWin.show();
   mainWin.init();
 

--- a/src/logClient/CentralWindow.cpp
+++ b/src/logClient/CentralWindow.cpp
@@ -97,7 +97,7 @@ CentralWindow::CentralWindow()
     log_frameNumber = new QLCDNumber(logControlWidget);
     log_frameNumber->setObjectName(QString::fromUtf8("log_frameNumber"));
     log_frameNumber->setGeometry(QRect(160, 90, 150, 25));
-    log_frameNumber->setNumDigits(7);
+    log_frameNumber->setDigitCount(7);
     log_frameNumber->setSegmentStyle(QLCDNumber::Flat);
     log_frameNumber->setProperty("intValue", QVariant(0));
     log_frame_forward = new QPushButton(logControlWidget);
@@ -108,21 +108,21 @@ CentralWindow::CentralWindow()
     log_totalFrames->setGeometry(QRect(480, 90, 150, 25));
     log_totalFrames->setAutoFillBackground(false);
     log_totalFrames->setSmallDecimalPoint(false);
-    log_totalFrames->setNumDigits(7);
+    log_totalFrames->setDigitCount(7);
     log_totalFrames->setSegmentStyle(QLCDNumber::Flat);
     logControl->setWidget(logControlWidget);
     this->addDockWidget(static_cast<Qt::DockWidgetArea>(8), logControl);
 
-    logControl->setWindowTitle(QApplication::translate("GuiControls", "LogControl", 0, QApplication::UnicodeUTF8));
-    log_backward->setText(QApplication::translate("GuiControls", "backward", 0, QApplication::UnicodeUTF8));
-    log_pause->setText(QApplication::translate("GuiControls", "pause", 0, QApplication::UnicodeUTF8));
-    log_forward->setText(QApplication::translate("GuiControls", "forward", 0, QApplication::UnicodeUTF8));
-    log_play->setText(QApplication::translate("GuiControls", "play", 0, QApplication::UnicodeUTF8));
-    log_slower->setText(QApplication::translate("GuiControls", "slower", 0, QApplication::UnicodeUTF8));
-    log_speed->setText(QApplication::translate("GuiControls", "Speed", 0, QApplication::UnicodeUTF8));
-    log_faster->setText(QApplication::translate("GuiControls", "faster", 0, QApplication::UnicodeUTF8));
-    log_frame_back->setText(QApplication::translate("GuiControls", "frame--", 0, QApplication::UnicodeUTF8));
-    log_frame_forward->setText(QApplication::translate("GuiControls", "frame++", 0, QApplication::UnicodeUTF8));
+    logControl->setWindowTitle(QApplication::translate("GuiControls", "LogControl", 0));
+    log_backward->setText(QApplication::translate("GuiControls", "backward", 0));
+    log_pause->setText(QApplication::translate("GuiControls", "pause", 0));
+    log_forward->setText(QApplication::translate("GuiControls", "forward", 0));
+    log_play->setText(QApplication::translate("GuiControls", "play", 0));
+    log_slower->setText(QApplication::translate("GuiControls", "slower", 0));
+    log_speed->setText(QApplication::translate("GuiControls", "Speed", 0));
+    log_faster->setText(QApplication::translate("GuiControls", "faster", 0));
+    log_frame_back->setText(QApplication::translate("GuiControls", "frame--", 0));
+    log_frame_forward->setText(QApplication::translate("GuiControls", "frame++", 0));
 
     logControl->hide();
     thread = new ViewUpdateThread(soccerView, drawMutex);

--- a/src/logClient/CentralWindow.h
+++ b/src/logClient/CentralWindow.h
@@ -23,13 +23,13 @@
 #define CENTRALWINDOW_H
 
 #include <QMainWindow>
-#include <QtGui/QGridLayout>
-#include <QtGui/QPushButton>
-#include <QtGui/QDockWidget>
-#include <QtGui/QWidget>
-#include <QtGui/QSlider>
-#include <QtGui/QLabel>
-#include <QtGui/QLCDNumber>
+#include <QGridLayout>
+#include <QPushButton>
+#include <QDockWidget>
+#include <QWidget>
+#include <QSlider>
+#include <QLabel>
+#include <QLCDNumber>
 #include "GraphicsPrimitives.h"
 #include "ClientThreading.h"
 

--- a/src/logClient/ClientThreading.cpp
+++ b/src/logClient/ClientThreading.cpp
@@ -179,7 +179,7 @@ int ViewUpdateThread::start_play_record()
         int last_slash = 0;
         for(int i = 0; i < fileName.length(); i++)
         {
-            if(fileName.at(i).toAscii() == (QChar('/')))
+            if(fileName.at(i).toLatin1() == (QChar('/')))
                 last_slash = i;
         }
         fileName.remove(last_slash, fileName.length()-last_slash);
@@ -187,13 +187,13 @@ int ViewUpdateThread::start_play_record()
 
     //What data shall I read?
     fileName = QFileDialog::getOpenFileName((QWidget*)this->parent(), tr("Open Logfile"), fileName, tr("Log Files (*.log)"));
-    std::cout << "fileName: " << fileName.toAscii().constData() << std::endl;
+    std::cout << "fileName: " << fileName.toLatin1().constData() << std::endl;
 
     // Read the existing log.
-    std::fstream input(fileName.toAscii().constData(), std::ios::in | std::ios::binary);
+    std::fstream input(fileName.toLatin1().constData(), std::ios::in | std::ios::binary);
     if (!input)
     {
-        std::cout << fileName.toAscii().constData() << ": File not found." << std::endl;
+        std::cout << fileName.toLatin1().constData() << ": File not found." << std::endl;
         return -1;
     }
     else if (!logs.ParseFromIstream(&input))

--- a/src/logClient/GraphicsPrimitives.h
+++ b/src/logClient/GraphicsPrimitives.h
@@ -32,7 +32,7 @@
 #include <QThread>
 #include <QGLWidget>
 #include <QMutex>
-#include <QtGui/QPushButton>
+#include <QPushButton>
 #include "field_default_constants.h"
 #include "robocup_ssl_client.h"
 #include "timer.h"

--- a/src/logClient/main.cpp
+++ b/src/logClient/main.cpp
@@ -23,6 +23,8 @@
 
 #include <stdio.h>
 #include <QTime>
+#include <QTextCodec>
+#include <QtGlobal>
 #include "CentralWindow.h"
 
 
@@ -38,6 +40,10 @@ int main(int argc, char *argv[])
 
 
     app = new QApplication(argc,argv);
+
+#if QT_MAJOR_VERSION == 4
+    QTextCodec::setCodecForTr(QTextCodec::codecForName("UTF-8"));
+#endif
 
     CentralWindow* centralWindow = new CentralWindow();
     centralWindow->show();

--- a/src/shared/util/qgetopt.cpp
+++ b/src/shared/util/qgetopt.cpp
@@ -95,9 +95,9 @@
 GetOpt::GetOpt()
 {
     if ( !qApp )
-	qFatal( "GetOpt: requires a QApplication instance to be constructed first" );
+    qFatal( "GetOpt: requires a QApplication instance to be constructed first" );
 
-    init( qApp->argc(), qApp->argv(), 1 );
+    init( qApp->arguments(), 1 );
 }
 
 /**
@@ -106,9 +106,9 @@ GetOpt::GetOpt()
 GetOpt::GetOpt( int offset )
 {
     if ( !qApp )
-	qFatal( "GetOpt: requires a QApplication instance to be constructed first" );
+    qFatal( "GetOpt: requires a QApplication instance to be constructed first" );
 
-    init( qApp->argc(), qApp->argv(), offset );
+    init( qApp->arguments(), offset );
 }
 
 /**
@@ -158,6 +158,22 @@ void GetOpt::init( int argc, char *argv[], int offset )
 	// arguments
 	for ( int i = offset; i < argc; ++i )
 	    args.append( QString::fromUtf8( argv[i] ) );
+    }
+}
+
+/**
+   \internal
+*/
+void GetOpt::init( const QStringList &arguments, int offset )
+{
+    numReqArgs = numOptArgs = 0;
+    currArg = 1; // appname is not part of the arguments
+    if (arguments.size() > 0) {
+        // application name
+        aname = QFileInfo(arguments.at(0)).fileName();
+        // arguments
+        for ( int i = offset; i < arguments.size(); ++i )
+            args.append(arguments.at(i));
     }
 }
 
@@ -275,7 +291,7 @@ bool GetOpt::parse( bool untilFirstSwitchOnly )
 	    }
 	    if ( t == LongOpt && opt.type == OUnknown ) {
 		if ( currOpt.type != OVarLen ) {
-		    qWarning( "Unknown option --%s",a.toAscii().constData() );
+            qWarning( "Unknown option --%s",a.toLatin1().constData() );
 		    return false;
 		} else {
 		    // VarLength options support arguments starting with '-'
@@ -368,7 +384,7 @@ bool GetOpt::parse( bool untilFirstSwitchOnly )
 	    } else {
 		QString n = currType == LongOpt ?
 			    currOpt.lname : QString( QChar( currOpt.sname ) );
-		qWarning( "Expected an argument after '%s' option", n.toAscii().constData() );
+        qWarning( "Expected an argument after '%s' option", n.toLatin1().constData() );
 		return false;
 	    }
 	    break;

--- a/src/shared/util/qgetopt.h
+++ b/src/shared/util/qgetopt.h
@@ -93,6 +93,7 @@ private:
     QMap<QString, int> setOptions;
 
     void init( int argc, char *argv[], int offset = 1 );
+    void init( const QStringList &arguments, int offset = 1 );
     void addOption( Option o );
     void setSwitch( const Option &o );
 

